### PR TITLE
Add decoder cache on top of fast beam search for transformer

### PR DESCRIPTION
**Pull request by francoishernandez**
This is an implementation of the transformer decoder cache.
As it is only beneficial for the fast beam search (thanks @guillaumekln!), its use is conditioned to the `-fast` option.

For 3k segments (WMT EN-DE 2014, sorted) on a GTX 1080:
BATCH 32 - BEAM 5
- fast beam: 77 seconds (peak VRAM usage 7.9G)
- fast beam + cache: 65 seconds (peak VRAM usage 2.4G)

BATCH 32 - BEAM 10
- fast beam: 137 seconds (peak VRAM usage 7.9G)
- fast beam + cache: 114 seconds (peak VRAM usage 3.7G)

### CODE REVIEW

File: onmt/decoders/decoder.py [hunk 1]
```diff
@@ -104,7 +104,7 @@ def __init__(self, rnn_type, bidirectional_encoder, num_layers,
         self._reuse_copy_attn = reuse_copy_attn
 
     def forward(self, tgt, memory_bank, state, memory_lengths=None,
-                step=None):
+                step=None, fast=None):
```
Comment (guillaumekln) [inline onmt/decoders/decoder.py]: Maybe @srush have a comment on adding this argument. For the context, it was apparently too costly to update the cache in `beam_update` so this feature expects translating with `-fast`.

If we go this way, all decoder classes should be updated. Related to https://github.com/OpenNMT/OpenNMT-py/pull/793.

File: onmt/decoders/transformer.py [hunk 2]
```diff
@@ -88,7 +91,10 @@ def forward(self, inputs, memory_bank, src_pad_mask, tgt_pad_mask,
 
         query_norm = self.layer_norm_2(query)
         mid, attn = self.context_attn(memory_bank, memory_bank, query_norm,
-                                      mask=src_pad_mask)
+                                      mask=src_pad_mask,
+                                      layer_cache=None,
```
Comment (guillaumekln) [inline onmt/decoders/transformer.py]: Why `None`? Looks like you need to update your numbers. :)

File: onmt/decoders/transformer.py [hunk 3]
```diff
@@ -255,28 +264,46 @@ def _all(self):
...
         Contains attributes that need to be updated in self.beam_update().
         """
-        return (self.previous_input, self.previous_layer_inputs, self.src)
+        if (self.previous_input is not None
+                and self.previous_layer_inputs is not None):
+            return (self.previous_input,
+                    self.previous_layer_inputs,
+                    self.src)
+        else:
+            return (self.src)
```
Comment (guillaumekln) [inline onmt/decoders/transformer.py]:
```python
return (self.src,)
```

File: onmt/decoders/transformer.py (hunk 2 above)
Comment (francoishernandez) [reply onmt/decoders/transformer.py]: Thanks... Just updated them. Hope this fits more your expectations :)

File: onmt/decoders/decoder.py (hunk 1 above)
Comment (srush) [reply onmt/decoders/decoder.py]: Hmm, yeah, I would really like these classes not to know about decoding, particularly in forward. If you have this option, maybe give it a more general now. Alternatively, you could have it be a flag on the module itself.

Comment (francoishernandez) [reply onmt/decoders/decoder.py]: Thank you both for your help. I added a `with_cache` argument to the `init_decoder_state` function, which is put to `True` only when going through `_fast_translate_batch`. Hope this fits better your expectations @srush !

**Pull request merged**